File: 1089.duplicate-zeros.java
class Solution {
    public void duplicateZeros(int[] arr) {
        /* find last element to copy in result */
        int r, w, len;
        for (r = len = 0; len < arr.length; r++) {
            if (arr[r] == 0)
                len += 2;
            else
                len++;
        }
        --r;

        /* handle edge case where a zero's duplicate is the first element to be thrown off */
        w = arr.length-1;
        if (len == arr.length + 1) {
            arr[w] = 0;
            w--;
            r--;
        }

        /* write the output with doubled zeroes into the array */
        while (r >= 0) {
            arr[w] = arr[r];
            w--;
            if (arr[r] == 0) {
                arr[w] = 0;
                w--;
            }
            r--;
        }
    }
}

File: 1089.duplicate-zeros.c
void duplicateZeros(int* arr, int arrSize) {
    int i, extras, r, w;

    /* find number of extra elements exceeding array space */
    for (i = extras = 0; i < arrSize; ++i) {
        if (arr[i] == 0)
            extras++;
    }

    /* find positions of read and write pointers */
    r = w = arrSize - 1;
    while (extras) {
        if (arr[r] == 0) {
            /* handle edge case when last zero's duplicate is first element to overflow the array */
            if (extras == 1) {
                arr[w--] = 0;
                extras--;
            }
            else
                extras -= 2;
        }
        else
            extras--;
        r--;
    }

    /* write transformed array with duplicate zeroes starting from the end */
    while(r >= 0) {
        arr[w--] = arr[r];
        if (arr[r] == 0)
            arr[w--] = 0;
        r--;
    }
}
